src/app/BufferedReadCallback.cpp

```diff
@@ -0,0 +1,256 @@
+
+void BufferedReadCallback::OnReportBegin(const ReadClient * apReadClient)
+{
+    mCallback.OnReportBegin(apReadClient);
+}
+
+void BufferedReadCallback::OnReportEnd(const ReadClient * apReadClient)
+{
+    CHIP_ERROR err = DispatchBufferedData(nullptr, mBufferedPath, StatusIB(), true);
+    if (err != CHIP_NO_ERROR)
+    {
+        mCallback.OnError(apReadClient, err);
+    }
+
+    mCallback.OnReportEnd(apReadClient);
+}
+
+CHIP_ERROR BufferedReadCallback::GenerateListTLV(TLV::ScopedBufferTLVReader & aReader)
+{
+    TLV::TLVType outerType;
+    Platform::ScopedMemoryBuffer<uint8_t> backingBuffer;
+
+    //
+    // To generate the final reconstituted list, we need to allocate a contiguous
+    // buffer than can hold the entirety of its contents. To do so, we need to figure out
+    // how big a buffer to allocate. This requires walking the buffered list items and computing their TLV sizes,
+    // sum theming all up and adding a bit of slop to account for the TLV array the list elements will go into.
+    //
+    // The alternative was to use a PacketBufferTLVWriter backed by chained packet buffers to
+    // write out the list - this would have removed the need for this first pass. However,
+    // we cannot actually back a TLVReader with a chained buffer since that violates the ability
+    // for us to create readers off-of readers. Each reader would assume exclusive ownership of the chained
+    // buffer and mutate the state within TLVPacketBufferBackingStore, preventing shared use.
+    //
+    // To avoid that, a single contiguous buffer is the best likely approach for now.
+    //
+    uint32_t totalBufSize = 0;
+    for (size_t i = 0; i < mBufferedList.size(); i++)
+    {
+        auto bufHandle = mBufferedList[i].Retain();
+        totalBufSize += bufHandle->TotalLength();
+    }
+
+    //
+    // Size of the start container and end container are just 1 byte each, but, let's just be safe.
+    //
+    totalBufSize += 4;
+
+    backingBuffer.Calloc(totalBufSize);
+    VerifyOrReturnError(backingBuffer.Get() != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    TLV::ScopedBufferTLVWriter writer(std::move(backingBuffer), totalBufSize);
+    ReturnErrorOnFailure(writer.StartContainer(TLV::AnonymousTag, TLV::kTLVType_Array, outerType));
+
+    for (auto & bufHandle : mBufferedList)
+    {
+        System::PacketBufferTLVReader reader;
+
+        reader.Init(std::move(bufHandle));
+
+        ReturnErrorOnFailure(reader.Next());
+        ReturnErrorOnFailure(writer.CopyElement(TLV::AnonymousTag, reader));
+    }
+
+    ReturnErrorOnFailure(writer.EndContainer(outerType));
+
+    writer.Finalize(backingBuffer);
+
+    aReader.Init(std::move(backingBuffer), totalBufSize);
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR BufferedReadCallback::AllocAndCopyElement(TLV::TLVReader & reader, System::PacketBufferHandle & handle)
+{
+    System::PacketBufferTLVWriter writer;
+    handle = System::PacketBufferHandle::New(chip::app::kMaxSecureSduLengthBytes);
```

Comment:
Again, seems like allocating this to the size the reader actually has to work with might be better.  We would still need the RightSize() call, but this allocation would typically be smaller...  At the very least, worth a comment for why we're knowingly over-allocating.

Of course the other thing we could do is to snapshot the reader, do a `Next()`, subtract the reader positions.  Followup for that if you think it's reasonable.

Reply:
I prototyped out the snapshotting solution, and the problem there is that because the reader's read point is positioned after the control octet + tag, the size retrieved is always going to under-shoot the actual size of the buffer (with the control octet + tag). I _could_ conservatively estimate the size of the control octet + tag size, but I think that's a little dangerous and breaks the abstraction here.

So for now, I'm just going to put a comment here explaining that, and in the future if this is a problem, we can figure out how to improve this.

Reply:
Thank you.  Having a comment explaining the pitfalls of what's not being done is very helpful here!